[Autocast] Fix edge case casting input directly to output

aboubezari opened the pull request:
## What does this PR do?

**Type of change:** Bug fix

**Overview:** If there a cast node connecting an input directly to an output, then the output will be totally disconnected due to naming issues. This fix will create specialized cast nodes for such edge cases and avoid removing them in the initial pass.

## Usage
Autocast precision converter

## Testing
Added a unittest that fails before my change, and passes after my fix.

## Before your PR is "*Ready for review*"
<!-- If you haven't finished some of the above items you can still open `Draft` PR. -->

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes
- **Did you write any new necessary tests?**: Yes
- **Did you add or update any necessary documentation?**: Yes
- **Did you update [Changelog](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CHANGELOG.rst)?**: No

gcunhase commented on tests/unit/onnx/autocast/test_precisionconverter.py:
```diff
@@ -1023,3 +1023,67 @@ def test_constant_cast_folding(model_with_constant_cast_patterns, low_precision_
+        value_info_map,
+        initializer_map,
+        node_to_init_map,
+        keep_io_types=True,
```
This test fails when `keep_io_types=False` due to a graph input becoming a graph output directly, which violates the assertion in ModelOpt that all original input and output names should be maintained in the quantized model.

galagam reviewed (commented):
This is indeed an edge case that challenges AutoCast's assumptions.
I'd like to propose an alternative approach: In GraphSanitizer, if an input is cast directly to output, inject an identity node.
@aboubezari @gcunhase

aboubezari commented on modelopt/onnx/autocast/graphsanitizer.py:
```diff
@@ -322,6 +323,33 @@ def _match_layernorm_pattern(self, mean_node: onnx.NodeProto) -> dict | None:
             logger.debug(f"Failed to match LayerNorm pattern at {mean_node.name}: {e!s}")
             return None
 
+    def sanitize_io_casts(self) -> None:
+        """Handle the special case where an input is casted directly to an output.
+
+        Inject an identity node after the cast node.
+        """
+        model_input_names = {input.name for input in self.model.graph.input}
+        model_output_names = {output.name for output in self.model.graph.output}
+        nodes_to_add = []
+        for node in self.model.graph.node:
+            if node.op_type == "Cast":
+                if node.input[0] in model_input_names and node.output[0] in model_output_names:
+                    cast_input_name = node.input[0]
+                    cast_output_name = node.output[0]
+                    cast_new_output_name = cast_input_name + "_io_cast_identity"
+                    nodes_to_add.append(
+                        helper.make_node(
+                            "Identity",
+                            inputs=[cast_new_output_name],
+                            outputs=[cast_output_name],
+                            name=node.name + "_io_cast_identity",
+                        )
+                    )
+                    node.output[0] = cast_new_output_name
+
+        for node in nodes_to_add:
+            self.model.graph.node.append(node)
```
Addressed the comment.

aboubezari commented:
> @aboubezari The PR is approved. Can you please rebase and sign your commits? See: https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CONTRIBUTING.md#%EF%B8%8F-signing-your-work
> 
> Once commits are signed and rebased, PR should be able to auto-merge.

@galagam I can confirm my commits are verified and signed off. I have just fetched upstream main and rebased.

galagam commented:
/ok to test 315ef04d4f9e4e0026fecc9eb8b5c0110852f8e4
